src/lists.py

```diff
@@ -204,9 +212,11 @@ def contains_tr(x: List[T], e: T) -> bool:
     >>> contains_tr(L(1, L(2, L(3, None))), 2)
     True
     """
+    return False if x.tail is None else x.head==e or contains_tr(x.tail,e) # is this tail recursion??! not too sure
```

Comment:
It _is_ tail recursive, or at least can be in languages that have the optimisation for it. The reason is a little subtle, though.

With logical `and` and `or`, you don't need both operands at the same time to determine the truth value. When you write `x.head==e or contains_tr(x.tail,e)` we can evaluate `x.head==e` first, and if it is true, we don't need to evaluate the recursive call to get the return value; we already know that the `or` expression is true. On the other hand, if `x.head != e`, then we don't need to remember that to perform the `or` operation later; the result of the `or` expression is the same as the result of the recursive call.

There is another feature to these logical operators that you are exploiting here. When you know that `x.head==e` is true, you don't evaluate the right-hand-side of the `or` expression. This is what you want; you don't want to explore the rest of the list when you have already found what you are searching for. For `or`, if the left-hand side is `True`, you don't evaluate the right-hand side. For `and`, you don't evaluate the right-hand side when the left-hand side is `False`. This is called "short-cutting" (for obvious reasons), and it is generally only something that happens for logical operators, and there only for built-in types. (If you define your own `or` and `and`, they will not short-cut in this way and you will evaluate both sides of the expression before you can evaluate).

Generally, you should assume that both sides of an expression are evaluated before you can evaluate the operation (or more general, all arguments to a function/method, which is what operators are under the hood). Some languages have so-called lazy evaluation where this isn't the case, R is one of them, but most languages behave this way. If that was true for `or`, then you wouldn't have tail recursion, and you would also continue the recursion after you had found the element you are looking for.

For logical operations on primitive data, though, you usually get short-cut operation, and then you have tail recursion.

That was a long answer, but it is a rather subtle issue. And you do have tail-recursion.